geoh5py/shared/cut_by_extent.py

```diff
@@ -0,0 +1,616 @@
+#  You should have received a copy of the GNU Lesser General Public License    '
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.           '
+# ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+from pydantic import BaseModel, ConfigDict, model_validator
+
+from geoh5py.shared.utils import clean_extent_for_intersection, normalize
+
+
+if TYPE_CHECKING:
+    from geoh5py.objects.geo_image import GeoImage
+    from geoh5py.objects.grid2d import Grid2D
+
+
+class Plane(BaseModel):
+    """
+    Represents a 3D plane with origin and basis vectors.
+
+    :param origin: Origin point of the plane coordinate system.
+    :param u_vector: U-axis basis vector.
+    :param v_vector: V-axis basis vector.
+    """
+
+    model_config = ConfigDict(frozen=True, arbitrary_types_allowed=True)
+
+    origin: np.ndarray
+    u_vector: np.ndarray
+    v_vector: np.ndarray
+
+    @model_validator(mode="after")
+    def validate_uv(self) -> Plane:
+        """
+        Ensure:
+            - origin, u_vector and v_vector are 3D.
+            - u_vector and v_vector are normalized.
+            - u and v are 3D and orthogonal.
+        """
+        if self.origin.shape != (3,):
+            raise ValueError(f"origin must be shape (3,), got {self.origin.shape}")
+        if self.u_vector.shape != (3,):
+            raise ValueError(f"u_vector must be shape (3,), got {self.u_vector.shape}")
+        if self.v_vector.shape != (3,):
+            raise ValueError(f"v_vector must be shape (3,), got {self.v_vector.shape}")
+        if not np.isclose(np.linalg.norm(self.u_vector), 1.0, atol=1e-6):
+            raise ValueError("u_vector is not normalized.")
+        if not np.isclose(np.linalg.norm(self.v_vector), 1.0, atol=1e-6):
+            raise ValueError("v_vector is not normalized.")
+        if not np.isclose(float(np.dot(self.u_vector, self.v_vector)), 0.0, atol=1e-8):
+            raise ValueError("u_vector and v_vector are not orthogonal.")
+
+        return self
+
+    @classmethod
+    def create_from_points(
+        cls,
+        origin: np.ndarray,
+        point_u: np.ndarray,
+        point_v: np.ndarray,
+    ) -> Plane:
+        """
+        Create a Plane from three points.
+
+        :param origin: Origin point of the plane.
+        :param point_u: Point defining the U direction.
+        :param point_v: Point defining the V direction.
+
+        :return: Plane object.
+        """
+        u_vector = normalize(
+            np.asarray(point_u, dtype=float) - np.asarray(origin, dtype=float)
+        )
+        v_vector = normalize(
+            np.asarray(point_v, dtype=float) - np.asarray(origin, dtype=float)
+        )
+
+        return cls(
+            origin=np.asarray(origin, dtype=float), u_vector=u_vector, v_vector=v_vector
+        )
+
+    @classmethod
```

Comment:
Unless it's a constructor, or you are modifying the class itself, just use a staticmethod